File: go/railroad_test.go
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

// railroad_test.go is the Go port of ts/test/railroad.test.js: node-level
// model construction, the text emitter, the single-node SVG/ASCII
// renderers, the plugin decoration, and error cases.
package tabnasrailroad

import (
	"errors"
	"reflect"
	"regexp"
	"strings"
	"testing"

	tabnas "github.com/tabnas/parser/go"
)

// ---- plugin load --------------------------------------------------

func TestPluginDecoratesInstance(t *testing.T) {
	tn := tabnas.Make()
	if Of(tn) == nil {
		t.Fatal("Of should never return nil")
	}
	// Before loading, the decoration is absent.
	if tn.Decoration(DecorationName) != nil {
		t.Fatal("railroad decoration should be absent before Plugin")
	}
	if err := Plugin(tn, nil); err != nil {
		t.Fatal(err)
	}
	api, ok := tn.Decoration(DecorationName).(*RailroadApi)
	if !ok || api == nil {
		t.Fatal("railroad decoration should be a *RailroadApi after Plugin")
	}
	// The API exposes the render helpers.
	if _, err := api.RenderNodeText(Terminal("x")); err != nil {
		t.Errorf("RenderNodeText: %v", err)
	}
}

func TestChildInstancesInheritDecoration(t *testing.T) {
	tn := tabnas.Make()
	if err := Plugin(tn, nil); err != nil {
		t.Fatal(err)
	}
	child, err := tn.Derive()
	if err != nil {
		t.Fatal(err)
	}
	if _, ok := child.Decoration(DecorationName).(*RailroadApi); !ok {
		t.Error("derived instance should inherit the railroad decoration")
	}
}

// ---- text emitter -------------------------------------------------

func TestTextEmitter(t *testing.T) {
	cases := []struct {
		node *RailroadNode
		want string
	}{
		{Terminal("hi"), `"hi"`},
		{NonTerminal("expr"), "expr"},
		{Sequence(Terminal("a"), Terminal("b")), `"a" "b"`},
		{MustChoice(Terminal("a"), Terminal("b")), `("a" | "b")`},
		{Optional(Terminal("a")), `["a"]`},
		{OneOrMore(Terminal("a"), nil), `"a"+`},
		{ZeroOrMore(Terminal("a"), nil), `{"a"}`},
	}
	for _, tc := range cases {
		got, err := ToText(tc.node)
		if err != nil {
			t.Errorf("ToText(%+v): %v", tc.node, err)
			continue
		}
		if got != tc.want {
			t.Errorf("ToText = %q, want %q", got, tc.want)
		}
	}
}

// ---- svg node renderer --------------------------------------------

func TestSvgNodeWellFormed(t *testing.T) {
	svg, err := RenderNodeSvg(Diagram(Sequence(Terminal("GET"), NonTerminal("path"))))
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasPrefix(svg, "<svg ") || !strings.HasSuffix(svg, "</svg>") {
		t.Errorf("svg not well-formed: %.40q...", svg)
	}
	w := svgAttr(t, svg, "width")
	h := svgAttr(t, svg, "height")
	if w <= 0 || h <= 0 {
		t.Errorf("width/height must be positive: %d/%d", w, h)
	}
	if !strings.Contains(svg, "GET") || !strings.Contains(svg, "path") || !strings.Contains(svg, "<rect") {
		t.Errorf("svg missing GET/path/<rect")
	}
}

func TestSvgNestedRendersWithoutError(t *testing.T) {
	node := Diagram(Sequence(
		Terminal("["),
		Optional(Sequence(NonTerminal("item"), ZeroOrMore(Sequence(Terminal(","), NonTerminal("item")), nil))),
		Terminal("]")))
	svg, err := RenderNodeSvg(node)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasPrefix(svg, "<svg ") {
		t.Errorf("svg not well-formed")
	}
}

func TestSvgStacksSequenceWithoutOverlap(t *testing.T) {
	svg, err := RenderNodeSvg(Sequence(Terminal("a"), Terminal("b"), Terminal("c")))
	if err != nil {
		t.Fatal(err)
	}
	re := regexp.MustCompile(`<rect[^>]*\sy="([\d.]+)"[^>]*\sheight="([\d.]+)"`)
	matches := re.FindAllStringSubmatch(svg, -1)
	if len(matches) != 3 {
		t.Fatalf("expected 3 rects, got %d", len(matches))
	}
	type band struct{ y, bottom float64 }
	bands := make([]band, len(matches))
	for i, m := range matches {
		y := atof(m[1])
		h := atof(m[2])
		bands[i] = band{y, y + h}
	}
	// Sort by y.
	for i := 1; i < len(bands); i++ {
		for j := i; j > 0 && bands[j].y < bands[j-1].y; j-- {
			bands[j], bands[j-1] = bands[j-1], bands[j]
		}
	}
	for i := 1; i < len(bands); i++ {
		if bands[i].y < bands[i-1].bottom {
			t.Errorf("sequence boxes overlap vertically: %v", bands)
		}
	}
}

// ---- ascii node renderer ------------------------------------------

func TestAsciiNodeSequence(t *testing.T) {
	out, err := RenderNodeAscii(Diagram(Sequence(Terminal("a"), NonTerminal("b"))))
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(out, `"a"`) || !strings.Contains(out, "b") {
		t.Errorf("ascii missing \"a\"/b:\n%s", out)
	}
}

func TestAsciiNodePlainPureAscii(t *testing.T) {
	out, err := RenderNodeAscii(MustChoice(Terminal("a"), Terminal("b")), AsciiOptions{Plain: true})
	if err != nil {
		t.Fatal(err)
	}
	for _, r := range out {
		if r > 127 {
			t.Fatalf("expected pure ASCII, found rune %q", r)
		}
	}
}

// ---- errors --------------------------------------------------------

func TestChoiceNoBranchesErrors(t *testing.T) {
	_, err := Choice()
	var re *RailroadError
	if !errors.As(err, &re) {
		t.Errorf("Choice() should return a *RailroadError, got %v", err)
	}
}

func TestUnknownNodeKindErrors(t *testing.T) {
	bogus := &RailroadNode{Kind: "bogus"}
	if _, err := RenderNodeSvg(bogus); err == nil {
		t.Errorf("RenderNodeSvg(bogus) should error")
	} else {
		var re *RailroadError
		if !errors.As(err, &re) {
			t.Errorf("expected *RailroadError, got %v", err)
		}
	}
	if _, err := ToText(bogus); err == nil {
		t.Errorf("ToText(bogus) should error")
	}
}

func TestInvalidNodeErrors(t *testing.T) {
	// A nil node is invalid (the Go analog of Sequence(null)).
	if _, err := norm(nil); err == nil {
		t.Errorf("norm(nil) should error")
	}
	if _, err := ToText(nil); err == nil {
		t.Errorf("ToText(nil) should error")
	}
}

// ---- whole-model rule ordering --------------------------------------

// The renderers must emit rules in the model's declared RuleOrder, hoisting
// Start to the front — NOT in alphabetical or Go map order. This pins the
// renderer half with a hand-built model whose order is deliberately neither
// alphabetical nor start-first; TestExtractionHonoursDeclarationOrder below
// pins the extraction half.
func TestRenderersHonourDeclaredRuleOrder(t *testing.T) {
	mk := func(s string) *RailroadNode {
		n, err := Norm(s)
		if err != nil {
			t.Fatal(err)
		}
		return n
	}
	model := &GrammarModel{
		Start: "mid",
		Rules: map[string]*RailroadNode{
			"zebra": mk("z"),
			"alpha": mk("a"),
			"mid":   mk("m"),
		},
		// Neither alphabetical nor start-first.
		RuleOrder: []string{"zebra", "alpha", "mid"},
	}
	want := []string{"mid", "zebra", "alpha"} // Start hoisted, rest in order.

	ascii, err := ModelToAscii(model)
	if err != nil {
		t.Fatal(err)
	}
	var got []string
	for _, line := range strings.Split(ascii, "\n") {
		if strings.HasSuffix(line, ":") && !strings.ContainsAny(line, " \t") {
			got = append(got, strings.TrimSuffix(line, ":"))
		}
	}
	if strings.Join(got, ",") != strings.Join(want, ",") {
		t.Errorf("ascii rule order = %v, want %v", got, want)
	}

	svg, err := ModelToSvg(model)
	if err != nil {
		t.Fatal(err)
	}
	var svgGot []string
	for _, m := range regexp.MustCompile(`<g id="([^"]+)">`).FindAllStringSubmatch(svg, -1) {
		svgGot = append(svgGot, m[1])
	}
	if strings.Join(svgGot, ",") != strings.Join(want, ",") {
		t.Errorf("svg rule order = %v, want %v", svgGot, want)
	}
}

// ExtractGrammar must put rules in the model in the order the GRAMMAR
// declared them, matching the TS side's `Object.keys(rsm)` walk — not the
// alphabetical order a Go map walk has to be sorted into. The engine
// reports it via (*Tabnas).RuleNames; extract.go's declaredUserRules is the
// consumer.
//
// A struct grammar states its order in GrammarSpec.RuleOrder (a Go map has
// none to recover); GrammarText fills the same field in from the source key
// order. The rule names here are chosen so declaration order is neither
// alphabetical nor its reverse, so a sort in either direction fails this.
func TestExtractionHonoursDeclarationOrder(t *testing.T) {
	declared := []string{"zebra", "alpha", "mid"}

	tn := tabnas.Make()
	err := tn.Grammar(&tabnas.GrammarSpec{
		V: 2,
		Rule: map[string]*tabnas.GrammarRuleSpec{
			"zebra": {Open: []*tabnas.GrammarAltSpec{{S: "#TX", P: "alpha"}}},
			"alpha": {Open: []*tabnas.GrammarAltSpec{{S: "#TX", P: "mid"}}},
			"mid":   {Open: []*tabnas.GrammarAltSpec{{S: "#TX"}}},
		},
		RuleOrder: declared,
	})
	if err != nil {
		t.Fatal(err)
	}

	model := ExtractGrammar(tn, &ExtractOptions{Start: "mid"})
	if got := model.RuleOrder; !reflect.DeepEqual(got, declared) {
		t.Errorf("RuleOrder = %v, want %v", got, declared)
	}

	// The last-resort entry rule (neither the caller nor cfg.rule.start names
	// one) is the FIRST rule declared, not the alphabetically first — the
	// same rule ts/extract.ts's `Object.keys(rsm).find(isUserRule)` picks.
	if got := firstUserRule(tn); got != "zebra" {
		t.Errorf("firstUserRule = %q, want %q", got, "zebra")
	}

	// The renderers then hoist Start and keep the rest in that order.
	ascii, err := ModelToAscii(model)
	if err != nil {
		t.Fatal(err)
	}
	var got []string
	for _, line := range strings.Split(ascii, "\n") {
		name := strings.TrimSuffix(line, ":")
		if strings.HasSuffix(line, ":") && !strings.ContainsAny(line, " \t") &&
			model.Rules[name] != nil {
			got = append(got, name)
		}
	}
	want := []string{"mid", "zebra", "alpha"}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("ascii rule order = %v, want %v", got, want)
	}
}

// ---- exported Norm / NodeEqual --------------------------------------

func TestNorm(t *testing.T) {
	// A bare string coerces to a Terminal, mirroring the TS norm export.
	n, err := Norm("(")
	if err != nil {
		t.Fatalf("Norm(string) errored: %v", err)
	}
	if !NodeEqual(n, Terminal("(")) {
		t.Errorf("Norm(\"(\") = %+v, want Terminal(\"(\")", n)
	}

	// A valid node passes through unchanged.
	nt := NonTerminal("expr")
	n, err = Norm(nt)
	if err != nil {
		t.Fatalf("Norm(node) errored: %v", err)
	}
	if n != nt {
		t.Errorf("Norm(node) should return the same node")
	}

	// nil, a typed-nil node, an empty-Kind node, and a non-Item value all error.
	for _, bad := range []any{nil, (*RailroadNode)(nil), &RailroadNode{}, 42} {
		if _, err := Norm(bad); err == nil {
			t.Errorf("Norm(%#v) should error", bad)
		} else {
			var re *RailroadError
			if !errors.As(err, &re) {
				t.Errorf("Norm(%#v): expected *RailroadError, got %v", bad, err)
			}
		}
	}
}

func TestNodeEqual(t *testing.T) {
	a := MustChoice(Terminal("a"), Optional(NonTerminal("b")),
		OneOrMore(Terminal("c"), Terminal(",")))
	b := MustChoice(Terminal("a"), Optional(NonTerminal("b")),
		OneOrMore(Terminal("c"), Terminal(",")))
	if !NodeEqual(a, b) {
		t.Errorf("structurally identical trees should be NodeEqual")
	}
	if !NodeEqual(SkipNode(), SkipNode()) {
		t.Errorf("NodeEqual(Skip, Skip) should be true")
	}
	if NodeEqual(Terminal("x"), NonTerminal("x")) {
		t.Errorf("differing kinds should not be NodeEqual")
	}
	if NodeEqual(Terminal("x"), Terminal("y")) {
		t.Errorf("differing text should not be NodeEqual")
	}
	if NodeEqual(a, MustChoice(Terminal("a"))) {
		t.Errorf("differing item counts should not be NodeEqual")
	}
	if NodeEqual(OneOrMore(Terminal("c"), Terminal(",")), OneOrMore(Terminal("c"), nil)) {
		t.Errorf("rep presence mismatch should not be NodeEqual")
	}
	if !NodeEqual(nil, nil) {
		t.Errorf("NodeEqual(nil, nil) should be true")
	}
	if NodeEqual(a, nil) || NodeEqual(nil, a) {
		t.Errorf("nil vs node should not be NodeEqual")
	}
}

// ---- helpers -------------------------------------------------------

func svgAttr(t *testing.T, svg, name string) int {
	t.Helper()
	key := name + `="`
	i := strings.Index(svg, key)
	if i < 0 {
		t.Fatalf("attr %q not found", name)
	}
	i += len(key)
	n := 0
	for i < len(svg) && svg[i] >= '0' && svg[i] <= '9' {
		n = n*10 + int(svg[i]-'0')
		i++
	}
	return n
}

func atof(s string) float64 {
	var whole, frac float64
	var div float64 = 1
	dot := false
	for _, c := range s {
		if c == '.' {
			dot = true
			continue
		}
		d := float64(c - '0')
		if dot {
			div *= 10
			frac += d / div
		} else {
			whole = whole*10 + d
		}
	}
	return whole + frac
}

File: rs/tests/railroad_test.rs
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

//! The port of `go/railroad_test.go` (itself the port of
//! `ts/test/railroad.test.js`): node-level model construction, the text
//! emitter, the single-node SVG and ASCII renderers, the plugin mark,
//! rule ordering, and the error cases.

mod common;

use regex::Regex;
use tabnas::{GrammarSpec, Tabnas};
use tabnas_railroad::{
    choice, diagram, extract_grammar, model_to_ascii, model_to_svg, node_equal, non_terminal,
    one_or_more, optional, render_node_ascii, render_node_svg, sequence, skip, terminal, to_text,
    zero_or_more, AsciiOptions, ExtractOptions, GrammarModel, RailroadDecoration, RailroadError,
    RailroadNode, SvgOptions, DECORATION_NAME,
};

// ---- plugin load ----------------------------------------------------

#[test]
fn plugin_marks_the_instance() {
    let mut parser = Tabnas::new();
    // Before loading, the mark is absent, and the API still binds.
    assert!(!tabnas_railroad::installed(&parser));
    assert_eq!(
        tabnas_railroad::of(&parser).render_node_text(&terminal("x")),
        "\"x\""
    );

    tabnas_railroad::railroad(&mut parser).expect("the plugin installs");
    assert!(tabnas_railroad::installed(&parser));
    assert_eq!(
        parser.decoration::<RailroadDecoration>(DECORATION_NAME),
        Some(&RailroadDecoration)
    );
}

#[test]
fn loads_through_use_plugin() {
    let mut parser = Tabnas::new();
    parser
        .use_plugin(tabnas_railroad::plugin(), None)
        .expect("the plugin installs");
    assert!(tabnas_railroad::installed(&parser));
}

#[test]
fn derived_instances_inherit_the_mark() {
    let mut parser = Tabnas::new();
    tabnas_railroad::railroad(&mut parser).expect("the plugin installs");
    let child = parser.derive(|_| {}).expect("the child derives");
    assert!(tabnas_railroad::installed(&child));
}

// ---- text emitter ---------------------------------------------------

#[test]
fn text_emitter() {
    let cases: Vec<(RailroadNode, &str)> = vec![
        (terminal("hi"), "\"hi\""),
        (non_terminal("expr"), "expr"),
        (sequence(["a", "b"]), "\"a\" \"b\""),
        (choice(["a", "b"]).unwrap(), "(\"a\" | \"b\")"),
        (optional("a"), "[\"a\"]"),
        (one_or_more("a", None), "\"a\"+"),
        (zero_or_more("a", None), "{\"a\"}"),
        (one_or_more("a", Some(terminal(","))), "\"a\"+ /* \",\" */"),
        (tabnas_railroad::comment("note"), "/* note */"),
        (skip(), ""),
    ];
    for (node, want) in cases {
        assert_eq!(to_text(&node), want, "to_text({node:?})");
    }
}

// ---- svg node renderer ----------------------------------------------

#[test]
fn svg_node_is_well_formed() {
    let svg = render_node_svg(
        &diagram([sequence([terminal("GET"), non_terminal("path")])]),
        &SvgOptions::default(),
    );
    assert!(svg.starts_with("<svg "), "svg not well-formed: {svg:.40}");
    assert!(svg.ends_with("</svg>"), "svg not well-formed");
    assert!(common::svg_attr(&svg, "width") > 0);
    assert!(common::svg_attr(&svg, "height") > 0);
    assert!(svg.contains("GET") && svg.contains("path") && svg.contains("<rect"));
}

#[test]
fn svg_nested_renders() {
    let node = diagram([sequence([
        terminal("["),
        optional(sequence([
            non_terminal("item"),
            zero_or_more(sequence([terminal(","), non_terminal("item")]), None),
        ])),
        terminal("]"),
    ])]);
    assert!(render_node_svg(&node, &SvgOptions::default()).starts_with("<svg "));
}

#[test]
fn svg_stacks_a_sequence_without_overlap() {
    let svg = render_node_svg(&sequence(["a", "b", "c"]), &SvgOptions::default());
    let re = Regex::new(r#"<rect[^>]*\sy="([\d.]+)"[^>]*\sheight="([\d.]+)""#).unwrap();
    let mut bands: Vec<(f64, f64)> = re
        .captures_iter(&svg)
        .map(|m| {
            let y: f64 = m[1].parse().unwrap();
            let h: f64 = m[2].parse().unwrap();
            (y, y + h)
        })
        .collect();
    assert_eq!(bands.len(), 3, "expected 3 rects");
    bands.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
    for pair in bands.windows(2) {
        assert!(
            pair[1].0 >= pair[0].1,
            "sequence boxes overlap vertically: {bands:?}"
        );
    }
}

#[test]
fn svg_links_nonterminals_through_link_for() {
    let opts = SvgOptions::with_link_for(|name| Some(format!("#rule-{name}")));
    let svg = render_node_svg(&non_terminal("expr"), &opts);
    assert!(svg.contains(r##"<a href="#rule-expr">"##));
    let plain = render_node_svg(&non_terminal("expr"), &SvgOptions::default());
    assert!(!plain.contains("<a "));
}

#[test]
fn svg_escapes_markup_in_labels() {
    let svg = render_node_svg(&terminal("<&>"), &SvgOptions::default());
    assert!(svg.contains("&lt;&amp;&gt;"));
    assert!(!svg.contains("<&>"));
}

// ---- ascii node renderer --------------------------------------------

#[test]
fn ascii_node_sequence() {
    let out = render_node_ascii(
        &diagram([sequence([terminal("a"), non_terminal("b")])]),
        &AsciiOptions::default(),
    );
    assert!(out.contains("\"a\"") && out.contains('b'), "{out}");
}

#[test]
fn ascii_plain_is_pure_ascii() {
    let out = render_node_ascii(&choice(["a", "b"]).unwrap(), &AsciiOptions::plain());
    assert!(out.is_ascii(), "expected pure ASCII, got {out}");
}

// ---- errors ---------------------------------------------------------

#[test]
fn choice_with_no_branches_errors() {
    let error = choice(Vec::<RailroadNode>::new()).unwrap_err();
    assert_eq!(
        error,
        RailroadError::new("railroad: choice needs at least one branch")
    );
    assert_eq!(
        error.to_string(),
        "railroad: choice needs at least one branch"
    );
}

#[test]
fn unknown_node_kind_is_rejected_on_decode() {
    // TypeScript rejects `{ kind: 'bogus' }` when it reaches a renderer;
    // a typed node cannot hold an unknown kind, so this port rejects it
    // when the JSON is decoded, before it reaches one.
    let error = RailroadNode::from_json(r#"{"kind":"bogus"}"#).unwrap_err();
    assert!(error.message.starts_with("railroad: invalid diagram node"));
    assert_eq!(error.node, Some(serde_json::json!({"kind":"bogus"})));
}

#[test]
fn invalid_node_values_are_rejected_on_decode() {
    // The analog of Sequence(null) and Norm(42): not a node at all.
    for bad in ["null", "42", "{}", "{\"text\":\"x\"}", "[]"] {
        let error = RailroadNode::from_json(bad).unwrap_err();
        assert!(
            error.message.starts_with("railroad: invalid diagram node"),
            "{bad}: {error}"
        );
    }
    assert!(GrammarModel::from_json("[]").is_err());
}

// ---- whole-model rule ordering --------------------------------------

/// The renderers must emit rules in the model's own order, hoisting
/// `start` to the front, never alphabetically. This pins the renderer
/// half with a hand-built model whose order is deliberately neither
/// alphabetical nor start-first.
#[test]
fn renderers_honour_declared_rule_order() {
    let mut model = GrammarModel {
        start: "mid".to_string(),
        ..Default::default()
    };
    model.rules.insert("zebra".to_string(), terminal("z"));
    model.rules.insert("alpha".to_string(), terminal("a"));
    model.rules.insert("mid".to_string(), terminal("m"));
    let want = ["mid", "zebra", "alpha"];

    assert_eq!(model.rule_order(), want);

    let ascii = model_to_ascii(&model, &AsciiOptions::default());
    let got: Vec<&str> = ascii
        .lines()
        .filter(|line| line.ends_with(':') && !line.contains([' ', '\t']))
        .map(|line| line.trim_end_matches(':'))
        .collect();
    assert_eq!(got, want, "ascii rule order");

    let svg = model_to_svg(&model);
    let re = Regex::new(r#"<g id="([^"]+)">"#).unwrap();
    let got: Vec<&str> = re
        .captures_iter(&svg)
        .map(|m| m.get(1).unwrap().as_str())
        .collect();
    assert_eq!(got, want, "svg rule order");
}

/// Extraction must put rules in the model in the order the GRAMMAR
/// declared them, matching the TypeScript `Object.keys(rsm)` walk. The
/// engine keeps insertion order; a serialized grammar states it in
/// `ruleOrder`. The names are chosen so declaration order is neither
/// alphabetical nor its reverse, so a sort in either direction fails.
#[test]
fn extraction_honours_declaration_order() {
    let declared = ["zebra", "alpha", "mid"];
    let mut parser = Tabnas::new();
    let spec = GrammarSpec::from_value(serde_json::json!({
        "v": 2,
        "rule": {
            "zebra": { "open": [ { "s": "#TX", "p": "alpha" } ] },
            "alpha": { "open": [ { "s": "#TX", "p": "mid" } ] },
            "mid":   { "open": [ { "s": "#TX" } ] },
        },
        "ruleOrder": declared,
    }))
    .expect("the grammar document is valid");
    parser.grammar(&spec).expect("the grammar installs");

    let model = extract_grammar(&parser, &ExtractOptions::with_start("mid"));
    let got: Vec<&String> = model.rules.keys().collect();
    assert_eq!(got, declared);
    assert_eq!(model.start, "mid");
    assert_eq!(model.rule_order(), ["mid", "zebra", "alpha"]);

    // The last-resort entry rule (neither the caller nor the options
    // name one) is the FIRST rule declared, not the alphabetically first.
    parser.options.rule.start = String::new();
    let model = extract_grammar(&parser, &ExtractOptions::default());
    assert_eq!(model.start, "zebra");

    let ascii = model_to_ascii(&model, &AsciiOptions::default());
    let got: Vec<&str> = ascii
        .lines()
        .filter(|line| line.ends_with(':') && !line.contains([' ', '\t']))
        .map(|line| line.trim_end_matches(':'))
        .filter(|name| model.rules.contains_key(*name))
        .collect();
    assert_eq!(got, ["zebra", "alpha", "mid"]);
}

// ---- norm and node_equal --------------------------------------------

#[test]
fn a_bare_string_is_a_terminal() {
    // The TypeScript `norm` coercion of its Item union.
    let node: RailroadNode = "(".into();
    assert!(node_equal(&node, &terminal("(")));
    let node: RailroadNode = String::from("expr").into();
    assert_eq!(node, terminal("expr"));
    assert_eq!(
        sequence(["(", "expr", ")"]),
        sequence([terminal("("), terminal("expr"), terminal(")")])
    );
}

#[test]
fn node_equal_is_structural() {
    let a = choice([
        terminal("a"),
        optional(non_terminal("b")),
        one_or_more(terminal("c"), Some(terminal(","))),
    ])
    .unwrap();
    let b = choice([
        terminal("a"),
        optional(non_terminal("b")),
        one_or_more(terminal("c"), Some(terminal(","))),
    ])
    .unwrap();
    assert!(node_equal(&a, &b));
    assert!(node_equal(&skip(), &skip()));
    assert!(!node_equal(&terminal("x"), &non_terminal("x")));
    assert!(!node_equal(&terminal("x"), &terminal("y")));
    assert!(!node_equal(&a, &choice([terminal("a")]).unwrap()));
    assert!(!node_equal(
        &one_or_more(terminal("c"), Some(terminal(","))),
        &one_or_more(terminal("c"), None)
    ));
}

#[test]
fn nodes_round_trip_through_json() {
    let node = diagram([sequence([
        terminal("["),
        optional(sequence([
            non_terminal("item"),
            zero_or_more(sequence([terminal(","), non_terminal("item")]), None),
        ])),
        terminal("]"),
        one_or_more(terminal("x"), Some(tabnas_railroad::comment("sep"))),
        skip(),
    ])]);
    let json = node.to_json();
    assert_eq!(RailroadNode::from_json(&json).unwrap(), node);
    // The shapes are the TypeScript object shapes exactly.
    assert_eq!(terminal("x").to_json(), r#"{"kind":"terminal","text":"x"}"#);
    assert_eq!(skip().to_json(), r#"{"kind":"skip"}"#);
    assert_eq!(
        one_or_more("x", None).to_json(),
        r#"{"kind":"oneOrMore","item":{"kind":"terminal","text":"x"}}"#
    );
    assert_eq!(
        RailroadNode::from_json(r#"{"kind":"oneOrMore","item":{"kind":"skip"},"rep":null}"#)
            .unwrap(),
        one_or_more(skip(), None)
    );
}

#[test]
fn accessors_read_each_shape() {
    assert_eq!(terminal("x").kind(), "terminal");
    assert_eq!(non_terminal("x").kind(), "nonterminal");
    assert_eq!(one_or_more("x", None).kind(), "oneOrMore");
    assert_eq!(zero_or_more("x", None).kind(), "zeroOrMore");
    assert_eq!(terminal("x").text(), Some("x"));
    assert_eq!(skip().text(), None);
    assert_eq!(sequence(["a"]).items().map(<[RailroadNode]>::len), Some(1));
    assert_eq!(optional("a").item(), Some(&terminal("a")));
    assert_eq!(
        one_or_more("a", Some(terminal(","))).rep(),
        Some(&terminal(","))
    );
    assert_eq!(one_or_more("a", None).rep(), None);
}

#[test]
fn version_is_exported() {
    assert_eq!(tabnas_railroad::VERSION, "0.3.6");
}
